xdsl/dialects/builtin.py

```diff
@@ -62,19 +132,20 @@ def from_string_attr(data: StringAttr) -> SymbolNameAttr:
 
 
 @irdl_attr_definition
-class FlatSymbolRefAttr(ParametrizedAttribute):
-    name = "flat_symbol_ref"
-    data: ParameterDef[StringAttr]
+class SymbolRefAttr(ParametrizedAttribute):
+    name = "symbol_ref"
+    root_reference: ParameterDef[StringAttr]
+    nested_references: ParameterDef[ArrayAttr[StringAttr]]
```

Comment:
How is this implemented in MLIR? To me, something that is called `nested` usually sounds like the nested ref should just be another symbolref, that has a name and can have one more nested symbolref. That's also what I would assume from the print and how I imagine you resolve a hierarchy of symboltables (which is what this fundamentally comes down to, AFAIU). 

I do understand tough that we can also "simulate" this behavior in a more efficient way by using an array an iterating over it. Just curious :)

Reply:
So I mimicked MLIR exactly here. I also believe historically it must have been a linked list-like structure, which got replaced by an array at one point for efficiency, maybe? In any case, this is how it is implemented in MLIR.

Reply:
The way `SymbolRefAttr` works in MLIR is the following:
```
root: StringAttr
nested: ArrayRef<FlatSymbolRefAttr>
```
However, `FlatSymbolRefAttr` is not actually an attribute. Rather, it is a class around `SymbolRefAttr` that also ensures that it has no nested references.

If we want to model that exactly, then yes we would need to add `SymbolRefAttr` as parameters of the `nested_references`, and possibly create a `FlatSymbolRefAttr` which is exactly `Annotated[SymbolRefAttr, IsSymbolFlatConstraint()]` or something.